File: conformance/src/conformance.test.ts
import { describe, it } from "node:test";
import assert from "node:assert/strict";
import { readdirSync, readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import { renderToStaticMarkup } from "react-dom/server";
import {
  applyTransition,
  createRegistry,
  createRenderer,
  type Transition,
  type View,
} from "@servewright/react";
import { registerShadcnPrimitives } from "@servewright/react-shadcn";

interface PrimitiveCase {
  primitive: string;
  root: View["root"];
  expect: { contains: string[] };
}

interface TransitionCase {
  name: string;
  initialView: View;
  transitions: Transition[];
  expect: { contains: string[] };
}

const casesDir = join(dirname(fileURLToPath(import.meta.url)), "..", "cases");
const registry = createRegistry();
registerShadcnPrimitives(registry);
const renderer = createRenderer(registry);

describe("conformance react", () => {
  for (const file of readdirSync(casesDir).filter((name) => name.endsWith(".json"))) {
    const raw = JSON.parse(readFileSync(join(casesDir, file), "utf8")) as PrimitiveCase | TransitionCase;

    if ("initialView" in raw) {
      it(`applies transition sequence (${file})`, () => {
        let view = raw.initialView;
        for (const transition of raw.transitions) {
          view = applyTransition(view, transition);
        }
        const markup = renderToStaticMarkup(renderer.render(view));
        for (const text of raw.expect.contains) {
          assert.match(markup, new RegExp(text.replace(/[.*+?^${}()|[\]\\]/g, "\\$&")));
        }
      });
      continue;
    }

    it(`renders ${raw.primitive} (${file})`, () => {
      const view: View = {
        servewrightVersion: "1.0",
        schemaVersion: "0.1.0",
        screen: "conformance",
        stateVersion: 0,
        root: raw.root,
      };

      const markup = renderToStaticMarkup(renderer.render(view));
      for (const text of raw.expect.contains) {
        assert.match(markup, new RegExp(text.replace(/[.*+?^${}()|[\]\\]/g, "\\$&")));
      }
    });
  }
});

File: dart/servewright_flutter/test/conformance_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:servewright_flutter/servewright_flutter.dart';
import 'package:servewright_flutter_material/servewright_flutter_material.dart';

void main() {
  final casesDir = Directory('../../conformance/cases');
  final registry = createRegistry();
  registerMaterialPrimitives(registry);
  final renderer = createRenderer(registry);

  for (final file in casesDir
      .listSync()
      .whereType<File>()
      .where((entry) => entry.path.endsWith('.json'))) {
    final testCase = jsonDecode(file.readAsStringSync()) as Map<String, dynamic>;
    final expectations = testCase['expect'] as Map<String, dynamic>;
    final expected = (expectations['contains'] as List<dynamic>?) ?? [];
    final notExpected = (expectations['notContains'] as List<dynamic>?) ?? [];
    final semanticsLabels = (expectations['semanticsLabels'] as List<dynamic>?) ?? [];
    final fileName = file.uri.pathSegments.last;

    if (testCase.containsKey('initialView')) {
      testWidgets('conformance flutter applies transition ($fileName)', (tester) async {
        var view = ServewrightView.fromJson(testCase['initialView'] as Map<String, dynamic>);
        for (final transitionJson in testCase['transitions'] as List<dynamic>) {
          final transition = ServewrightTransition.fromJson(transitionJson as Map<String, dynamic>);
          view = TransitionApplier.apply(view, transition);
        }

        await tester.pumpWidget(
          MaterialApp(home: Scaffold(body: renderer.render(view))),
        );

        for (final text in expected) {
          expect(find.textContaining('$text'), findsWidgets);
        }
        for (final text in notExpected) {
          expect(find.text('$text'), findsNothing);
        }
        for (final label in semanticsLabels) {
          expect(find.bySemanticsLabel('$label'), findsWidgets);
        }
      });
      continue;
    }

    final primitive = testCase['primitive'] as String;
    final rootJson = testCase['root'] as Map<String, dynamic>;

    testWidgets('conformance flutter renders $primitive ($fileName)', (tester) async {
      final view = ServewrightView(
        servewrightVersion: '1.0',
        schemaVersion: testCase['schemaVersion'] as String? ?? '0.1.0',
        screen: 'conformance',
        stateVersion: 0,
        root: ServewrightNode.fromJson(rootJson),
      );

      await tester.pumpWidget(
        MaterialApp(home: Scaffold(body: renderer.render(view))),
      );

      for (final text in expected) {
        expect(find.textContaining('$text'), findsWidgets);
      }
      for (final text in notExpected) {
        expect(find.text('$text'), findsNothing);
      }
      for (final label in semanticsLabels) {
        expect(find.bySemanticsLabel('$label'), findsWidgets);
      }
    });
  }
}
